### PR TITLE
Add M141 Heated Chamber support

### PR DESCRIPTION
**Pull request by 3dlabsio**
Description
#########

This is a proposed implementation of actively heated chamber support, with a corresponding M141 command. M141 seems to be the standard gcode for controlling chamber temperature on other firmware.

I built upon the temp chamber support that was merged in previously. This has been tested and works on my Creatbot D600 Pro. M191, and PID support still needs to be added. For now this only supports bang-bang.

Benefits
##########

Adds support for an active chamber heater to Marlin.

Related Issues
#########
Fulfills a previous request for M141 implementation.

### CODE REVIEW

**Comment by silentninja1** [comment]
This has been on my todo list for awhile, but very low on the list because im not sure how an implementation here would interact with the stratasys patents. Regardless, it has been a requested feature and temp monitoring and fan control have been in place for awhile. Ill definitely look over some of this after the day job today regardless, but the call on what to do with it isnt mine :)

**Comment by 3dlabsio** [comment]
Sounds good. We had to do the work to get it going on our machine, and wanted to share that. I can see the concern with patents. It has been incorporated into other open source firmware if that helps: https://github.com/dc42/RepRapFirmware/blob/dev/src/Heating/Heat.cpp

I think a violation boils down to the structure of your machine as described in the patent. I.E. placing your motors outside of the heated zone. But we'll let you guys decide :)

**Comment by InsanityAutomation** [comment]
Yeah, my original concept was expanding the e3d water cooled hotend to the steppers to avoid that part!

**Comment by 3dlabsio** [comment]
https://patents.google.com/patent/US6722872

"the x-y-z gantry is located external to the build chamber and is separated from the chamber by a deformable thermal insulator"

**Comment by InsanityAutomation** [comment]
Iirc they have 3 different ones, but I'm not an expert on it. All along those lines.

File: Marlin/src/module/temperature.cpp [hunk 1]
```diff
@@ -911,6 +952,46 @@ void Temperature::manage_heater() {
       #endif
     }
   #endif // HAS_HEATED_BED
+
+	#if HAS_TEMP_CHAMBER
```
Comment (thinkyhead) [inline Marlin/src/module/temperature.cpp]: Two spaces for indentation. No tabs!

File: Marlin/src/module/temperature.cpp [hunk 2]
```diff
@@ -911,6 +952,46 @@ void Temperature::manage_heater() {
...
+            _temp_error(-2, PSTR(MSG_T_HEATING_FAILED), TEMP_ERR_PSTR(MSG_HEATING_FAILED_LCD, -2));
+          else                                                            // Start again if the target is still far off
+            start_watching_chamber();
+        }
+      #endif // WATCH_THE_CHAMBER
+
+      if (PENDING(ms, next_chamber_check_ms)) return;
+      next_chamber_check_ms = ms + CHAMBER_CHECK_INTERVAL;
+      
+      if(WITHIN(current_temperature_chamber, CHAMBER_MINTEMP, CHAMBER_MAXTEMP)){
```
Comment (thinkyhead) [inline Marlin/src/module/temperature.cpp]: A single space is required between `if` and the opening parenthesis.

File: Marlin/src/module/temperature.cpp [hunk 3]
```diff
@@ -911,6 +952,46 @@ void Temperature::manage_heater() {
...
+        #if ENABLED(CHAMBER_LIMIT_SWITCHING)
+          if (current_temperature_chamber >= target_temperature_chamber + CHAMBER_HYSTERESIS)
+            soft_pwm_amount_chamber = 0;
+          else if (current_temperature_chamber <= target_temperature_chamber - (CHAMBER_HYSTERESIS))
+            soft_pwm_amount_chamber = MAX_CHAMBER_POWER >> 1;
+        #else // !PIDTEMPCHAMBER && !CHAMBER_LIMIT_SWITCHING
+          soft_pwm_amount_chamber = current_temperature_chamber < target_temperature_chamber ? MAX_CHAMBER_POWER >> 1 : 0;
+        #endif
+      } else {
```
Comment (thinkyhead) [inline Marlin/src/module/temperature.cpp]: The correct format for `else` is:
```cpp
}
else {
```

File: Marlin/src/module/temperature.cpp [hunk 4]
```diff
@@ -1489,7 +1614,7 @@ void Temperature::init() {
           SERIAL_ECHOPAIR(" ;  Idle Timeout:", bed_idle_timeout_exceeded);
         SERIAL_EOL();
     */
-
+
```
Comment (thinkyhead) [inline Marlin/src/module/temperature.cpp]: Please be wary of trailing whitespace.

File: Marlin/src/module/temperature.cpp [hunk 5]
```diff
@@ -573,12 +591,26 @@ int Temperature::getHeaterPower(const int heater) {
...
 void Temperature::_temp_error(const int8_t e, PGM_P const serial_msg, PGM_P const lcd_msg) {
   static bool killed = false;
+  /*
   if (IsRunning()) {
     SERIAL_ERROR_START();
     serialprintPGM(serial_msg);
     SERIAL_ERRORPGM(MSG_STOPPED_HEATER);
     if (e >= 0) SERIAL_ERRORLN((int)e); else SERIAL_ERRORLNPGM(MSG_HEATER_BED);
   }
+  */
```
Comment (thinkyhead) [inline Marlin/src/module/temperature.cpp]: No need to comment out old code that's being replaced. Just replace it. We have a version control system to see the changes.

File: Marlin/src/gcode/gcode.cpp [hunk 6]
```diff
@@ -376,6 +376,11 @@ void GcodeSuite::process_parsed_command(
         case 190: M190(); break;                                  // M190: Wait for bed temperature to reach target
       #endif
 
+      #if ENABLED(HAS_HEATED_CHAMBER)
```
Comment (thinkyhead) [inline Marlin/src/gcode/gcode.cpp]: This option has not been added to any `Configuration.h` file. Note that the naming should not include `HAS_` for configurable options, but only for items defined in the `Conditionals*` files. And `HAS_` items defined in the `Conditionals*` files should not use `ENABLED()`. I would assume that this feature is enabled by having a temp sensor defined for the heated chamber.

**Comment by thinkyhead** [comment]
It's a good start, but needs some changes for completion and to comply with Marlin coding standards.

**Comment by 3dlabsio** [comment]
> 
> 
> It's a good start, but needs some changes for completion and to comply with Marlin coding standards.

Thanks for the feedback - I'll get the formatting changes cleaned up and be sure to adhere to those guidelines better going forward. We'll also refactor the other things you pointed out. 

I do realize that a Configuration.h option will need to be added as well as a Configuration_adv.h but wasn't sure about the best way to do that. I see several in the examples sub folders as well as a "default" set.  Shall we just add the commented out options required to the main Configuration/_adv.h files in line with the others?

File: Marlin/src/gcode/gcode.cpp (hunk 6 above)
Comment (3dlabsio) [reply Marlin/src/gcode/gcode.cpp]: Yes, you are correct that is it is enabled by having a temp sensor defined. This is meant to be an additional option on top of that, nested within the temp_chamber option. What do you suggest for a name?

**Comment by thinkyhead** [comment]
Typically, new options are added to the appropriate place in either `Configuration.h` or `Configuration_adv.h`, and then those same changes are propagated to all the examples. New features are disabled by default. It's easier to review changes if the example configuration updates are done in a separate commit.

**Comment by mb300sd** [comment]
I spent some time porting this to 1.1.x for my printer, now that I took another look, it might have been better to port the Raise3D changes to 2.0, but it's done.

I fixed some of the issues, and added entries to configuration, so it might help here.

https://github.com/mb300sd/Marlin-Raise3D-N2-Plus/commit/72041e2106b8ee1ce2a5f5c496b5551619485fd8

**Comment by InsanityAutomation** [comment]
Does your 1.1 r3d branch support the touchscreen? At the moment I haven't made an attempt to get that working, but I have a 2.0 branch setup for it.

**Comment by mb300sd** [comment]
> Does your 1.1 r3d branch support the touchscreen? At the moment I haven't made an attempt to get that working, but I have a 2.0 branch setup for it.

Yeah, it's based off of https://github.com/ABH10/Marlin-for-Raise3D-N-series/tree/1.1.x_Raise3D

**Comment by 3dlabsio** [comment]
> 
> 
> I spent some time porting this to 1.1.x for my printer, now that I took another look, it might have been better to port the Raise3D changes to 2.0, but it's done.
> 
> I fixed some of the issues, and added entries to configuration, so it might help here.
> 
> [mb300sd/Marlin-Raise3D-N2-Plus@72041e2](https://github.com/mb300sd/Marlin-Raise3D-N2-Plus/commit/72041e2106b8ee1ce2a5f5c496b5551619485fd8)

Sorry for the long period of inactivity on this. I've been busy working on other projects.

So we actually ended up implementing this in a 1.1.9 branch of another fork, but the core functionality is there. We did a massive cleanup of all the things suggested but haven't gotten around to implementing in 2.x. 

The branch with the initial PR here, and subsequent PR with fixes is here: 
https://github.com/3dlabsio/ADVi3pp-Marlin/tree/advi-creatbot-d600-dev

I'll try to get this updated for 2.x in the new year, unless someone else would like to jump on it sooner.

**Comment by thinkyhead** [comment]
I've squashed, rebased, and cleaned up the branch a bit. To get these changes into your working copy use the Git Console and the following commands:
```
git fetch origin
git checkout bugfix-2.0.x-M141-heated-chamber-support
git reset --hard origin/bugfix-2.0.x-M141-heated-chamber-support
```

File: Marlin/src/gcode/gcode.cpp (hunk 6 above)
Comment (thinkyhead) [reply Marlin/src/gcode/gcode.cpp]: See the commits in their current form.

**Comment by thinkyhead** [comment]
Added a chamber heating menu item, reorganized the temperature code into logical collections of data, and other enhancements…

**Comment by 3dlabsio** [comment]
Thanks for picking this up guys. Glad to see it available in the main branch now.

**Comment by nicksears** [comment]
wondering if M141 should work yet?

I'm using Marlin-bugfix-2.0.x CubeX from 8/28/19 and no luck

I want to use a (bang-bang) heater to get the chamber up to ~70C (https://www.amazon.com/dp/B003XDTWN2) not the opposite (auto fan pin and temp like with extruders)

I've done the following to try and enable temp control of a heated chamber:  

in config:
#define TEMP_SENSOR_CHAMBER 5 //(I think thats it)

and in config_adv:
  #define HEATER_CHAMBER_PIN       5//44   // using extra servo pin to drive ssr

Again, I'm not changing chamber auto fan pin/temp because I dont want to use it that way

any thoughts? I've seen many other threads, but this is most recent one still open...I feel like this should be easy to enable.

I know many others have suggested to use an extruder, but without the ability to enable/disable thermal protection on individual hotends this wont work (or set individual hysteresis, etc)

**Comment by thinkyhead** [comment]
This Issue Queue is for Marlin bug reports and development-related issues, and we prefer not to handle user-support questions here. (As noted on [this page](https://github.com/MarlinFirmware/Marlin/blob/1.1.x/.github/contributing.md#i-dont-want-to-read-this-whole-thing-i-just-have-a-question).) For best results getting help with configuration and troubleshooting, please use the following resources:

- MarlinFW.org [Marlin Documentation](http://marlinfw.org)
- RepRap.org [Marlin Forum](http://forums.reprap.org/list.php?415)
- [Tom's 3D Forums](https://discuss.toms3d.org/)
- Facebook Group ["Marlin Firmware"](https://www.facebook.com/groups/1049718498464482/)
- Facebook Group ["Marlin Firmware for 3D Printers"](https://www.facebook.com/groups/3Dtechtalk/)
- [Marlin Configuration](https://www.youtube.com/results?search_query=marlin+configuration) on YouTube
- Marlin Discord server. Join link: https://discord.gg/n5NJ59y

After seeking help from the community, if the consensus points to to a bug in Marlin, then you should post a [bug report](https://github.com/MarlinFirmware/Marlin/issues/new/choose).

**Comment by biocrypto730** [comment]
Any clue why this works on my MKS BASE (8-bit) but not my MKS SGEN? With roughly the same code : https://github.com/MarlinFirmware/Marlin/issues/15437

**Comment by livello** [comment]
@3dlabsio Hi there! Can you share your experience with creatbot d600 pro and marlin firmware? Is it usable or we should stay with factory Chinese closed fork?
Can you specify what type of heater original temperature sensor is used in d600 pro?

**Comment by thisiskeithb** [comment]
@livello:

This Issue Queue is for Marlin bug reports and development-related issues, and we prefer not to handle user-support questions here. (As noted on [this page](https://github.com/MarlinFirmware/Marlin/blob/bugfix-2.1.x/.github/contributing.md#i-dont-want-to-read-this-whole-thing-i-just-have-a-question).) For best results getting help with configuration and troubleshooting, please use the following resources:

- [Marlin Documentation](https://marlinfw.org) hosted on MarlinFW.org
- [Marlin Forum](https://forums.reprap.org/list.php?415) hosted on RepRap.org
- [Marlin Firmware](https://www.facebook.com/groups/1049718498464482/) Facebook Group 
- [Marlin Firmware for 3D Printers](https://www.facebook.com/groups/3Dtechtalk/) Facebook Group
- [Marlin Configuration](https://www.youtube.com/results?search_query=marlin+configuration) on YouTube
- [Marlin Firmware](https://discord.gg/n5NJ59y) Discord server

After seeking help from the community, if the consensus points to a bug in Marlin, then you should post a [bug report](https://github.com/MarlinFirmware/Marlin/issues/new/choose).

**Pull request merged**